src/ra_server_proc.erl

```diff
@@ -1188,6 +1230,81 @@ handle_effect(RaftState, {aux, Cmd}, EventType, State0, Actions0) ->
                        State0#state{server_state = ServerState}),
     {State, Actions0 ++ Actions};
+handle_effect(_RaftState, {add_member, Conf, ServerId, Members}, _EventType,
+              #state{server_state = SS, monitors = Monitors} = State, Actions) ->
+    #{name := System} = ra_server:system_config(SS),
+    #{member_eval_pid := OldPid} = SS,
+    Me = self(),
+    %% Check if an member_eval pid is already ongoing. Do we need to monitor the pid?
+    Pid = case is_pid(OldPid) andalso is_process_alive(OldPid) of
+              true ->
+                  %% There is already a process working, do nothing
+                  OldPid;
+              false ->
+                  %% No process working, safe to spawn a new one.
+                  spawn(fun() ->
+                                Res = case ra:start_server(System, Conf) of
+                                          ok ->
+                                              case ra:add_member(Members, ServerId) of
+                                                  {ok, _, _} = R ->
+                                                      R;
+                                                  {timeout, _} ->
+                                                      ra:force_delete_server(System, ServerId),
+                                                      ra:remove_member(Members, ServerId),
+                                                      {error, timeout};
+                                                  E ->
+                                                      ra:force_delete_server(System, ServerId),
+                                                      E
+                                              end;
+                                          E ->
+                                              E
+                                      end,
+                                ok = gen_statem:cast(Me, {add_member, ServerId, Res})
+                        end)
+          end,
+    %% Q: Storing the Pid, but is monitoring needed, or just check if
+    %% process is alive in each lap?
+    {State#state{monitors = ra_monitors:add(Pid, member_eval, Monitors),
+                 server_state = SS#{member_eval_pid => Pid}}, Actions};
+handle_effect(_RaftState, {remove_member, ServerId, Members}, _EventType,
+              #state{server_state = SS, monitors = Monitors} = State, Actions) ->
+    #{name := System} = ra_server:system_config(SS),
+    #{member_eval_pid := OldPid} = SS,
+    Me = self(),
+    %% Check if an member_eval pid is already ongoing. Do we need to monitor the pid?
+    Pid = case is_pid(OldPid) andalso is_process_alive(OldPid) of
+              true ->
+                  %% There is already a process working, do nothing
+                  OldPid;
+              false ->
+                  %% No process working, safe to spawn a new one.
+                  spawn(fun() ->
+                                Res = case ra:remove_member(Members, ServerId) of
```

Comment:
I think these should only be handled by the `leader` state, if they ended up being triggered on a follower they should just be ignored. Also the first thing the temporary process should do is a consistent query into the leader just to check it is an "active" leader and not a partitioned one. `ra:remove_member/2` returns after the command has been added to the log, _not_ after consensus so it isn't quite enough.

Reply:
Got it. How do I make sure its an active leader? Call to `leader_id` enough?